File: backend/app/services/delivery/telegram.py
"""Telegram outbound delivery with prepare / call / finalize.

Telegram Bot API has no real provider-side idempotency for ``sendMessage``;
``X-IMMIGROME-Idempotency-Key`` is local-only.

**At-most-once automatic delivery:** the outbox worker commits a durable
``DeliveryAttempt`` in ``PROCESSING`` *before* any HTTP call. After a
successful send, finalize marks SUCCESS. If the process crashes after send
but before finalize, reclaim finds ``PROCESSING`` (or
``UNKNOWN_REQUIRES_REVIEW``) and must **not** auto-resend — it leaves /
ensures ``UNKNOWN_REQUIRES_REVIEW`` for human confirmation.

Ambiguous transport/5xx errors during the call also finalize as
``UNKNOWN_REQUIRES_REVIEW``. Definite client errors fail the attempt and may
retry the outbox event (ordinary retry). Hermes/Calendar use separate
idempotent recovery; Telegram does not automatically resend after ambiguity.
"""

from __future__ import annotations

from dataclasses import dataclass
from datetime import UTC, datetime
from typing import Any, Literal
from uuid import uuid4

import httpx
from sqlalchemy import func, insert, select, update
from sqlalchemy.ext.asyncio import AsyncSession

from app.channels.telegram import TelegramAdapter
from app.core.config import get_settings
from app.db.tables import (
    channel_identity_table,
    conversation_message_table,
    conversation_table,
    delivery_attempt_table,
)
from app.observability.metrics import incr

AMBIGUOUS_TELEGRAM_ERRORS = (
    httpx.TimeoutException,
    httpx.ConnectError,
    httpx.RemoteProtocolError,
    httpx.ReadError,
    httpx.WriteError,
    httpx.NetworkError,
)

TelegramPrepareAction = Literal["send", "skip_success", "skip_unknown"]


@dataclass(frozen=True)
class PreparedTelegramDelivery:
    action: TelegramPrepareAction
    message_id: str
    conversation_id: str | None
    chat_id: str | None
    body: str | None
    attempt_id: str | None
    idempotency_key: str


@dataclass(frozen=True)
class TelegramCallResult:
    provider_message_id: str


def _is_ambiguous_after_send_start(error: Exception) -> bool:
    if isinstance(error, AMBIGUOUS_TELEGRAM_ERRORS):
        return True
    if isinstance(error, httpx.HTTPStatusError):
        return error.response is not None and error.response.status_code >= 500
    return False


async def _mark_message_unknown(
    db: AsyncSession, message_id: str, *, now: datetime
) -> None:
    await db.execute(
        update(conversation_message_table)
        .where(conversation_message_table.c.id == message_id)
        .values(deliveryStatus="UNKNOWN_REQUIRES_REVIEW", updatedAt=now)
    )


async def _ensure_unknown_attempt(
    db: AsyncSession,
    *,
    message_id: str,
    outbox_event_id: str,
    error_code: str,
    error_message: str,
    attempt_id: str | None = None,
) -> None:
    now = datetime.now(UTC)
    if attempt_id:
        await db.execute(
            update(delivery_attempt_table)
            .where(delivery_attempt_table.c.id == attempt_id)
            .values(
                status="UNKNOWN_REQUIRES_REVIEW",
                errorCode=error_code,
                errorMessage=error_message[:1000],
            )
        )
    else:
        processing = (
            await db.execute(
                select(delivery_attempt_table).where(
                    delivery_attempt_table.c.messageId == message_id,
                    delivery_attempt_table.c.provider == "TELEGRAM",
                    delivery_attempt_table.c.status == "PROCESSING",
                )
            )
        ).mappings().all()
        for row in processing:
            await db.execute(
                update(delivery_attempt_table)
                .where(delivery_attempt_table.c.id == row.id)
                .values(
                    status="UNKNOWN_REQUIRES_REVIEW",
                    errorCode=error_code,
                    errorMessage=error_message[:1000],
                )
            )
        if not processing:
            existing_unknown = (
                await db.execute(
                    select(delivery_attempt_table.c.id).where(
                        delivery_attempt_table.c.messageId == message_id,
                        delivery_attempt_table.c.provider == "TELEGRAM",
                        delivery_attempt_table.c.status == "UNKNOWN_REQUIRES_REVIEW",
                    )
                )
            ).scalar_one_or_none()
            if existing_unknown is None:
                attempt = (
                    await db.execute(
                        select(func.count())
                        .select_from(delivery_attempt_table)
                        .where(
                            delivery_attempt_table.c.messageId == message_id,
                            delivery_attempt_table.c.provider == "TELEGRAM",
                        )
                    )
                ).scalar_one() + 1
                await db.execute(
                    insert(delivery_attempt_table).values(
                        id=uuid4().hex,
                        messageId=message_id,
                        outboxEventId=outbox_event_id,
                        provider="TELEGRAM",
                        attempt=attempt,
                        status="UNKNOWN_REQUIRES_REVIEW",
                        errorCode=error_code,
                        errorMessage=error_message[:1000],
                        createdAt=now,
                    )
                )
    await _mark_message_unknown(db, message_id, now=now)
    incr("unknown_delivery", provider="TELEGRAM")


async def prepare_telegram_delivery(
    db: AsyncSession, event: dict[str, Any]
) -> PreparedTelegramDelivery:
    """Short DB transaction: load message, fence priors, insert PROCESSING attempt.

    Must COMMIT before ``call_telegram_delivery``. Does not perform HTTP.
    """
    message_id = str(event["payloadJson"].get("message_id") or "")
    if not message_id:
        raise ValueError("message.send.requested.v1 requires message_id")
    row = (
        await db.execute(
            select(
                conversation_message_table,
                conversation_table.c.leadId,
                conversation_table.c.studentId,
                conversation_table.c.automationPausedAt,
            )
            .join(
                conversation_table,
                conversation_table.c.id == conversation_message_table.c.conversationId,
            )
            .where(conversation_message_table.c.id == message_id)
        )
    ).mappings().first()
    if row is None:
        raise ValueError("Outbound message not found")
    if row.direction != "OUTBOUND":
        raise ValueError("Only outbound messages can be delivered")
    if row.policyStatus not in {"ALLOWED", "APPROVED"}:
        raise ValueError("Outbound message has not passed policy")
    if row.automationPausedAt is not None:
        raise ValueError("Conversation automation is paused")
    if not row.body:
        raise ValueError("Telegram text message is empty")

    prior_success = (
        await db.execute(
            select(delivery_attempt_table.c.providerResponseId).where(
                delivery_attempt_table.c.messageId == message_id,
                delivery_attempt_table.c.provider == "TELEGRAM",
                delivery_attempt_table.c.status == "SUCCESS",
            )
        )
    ).scalar_one_or_none()
    if prior_success is not None:
        return PreparedTelegramDelivery(
            action="skip_success",
            message_id=message_id,
            conversation_id=row.conversationId,
            chat_id=None,
            body=None,
            attempt_id=None,
            idempotency_key=event["idempotencyKey"],
        )

    prior_ambiguous = (
        await db.execute(
            select(delivery_attempt_table.c.id, delivery_attempt_table.c.status).where(
                delivery_attempt_table.c.messageId == message_id,
                delivery_attempt_table.c.provider == "TELEGRAM",
                delivery_attempt_table.c.status.in_(
                    ("PROCESSING", "UNKNOWN_REQUIRES_REVIEW")
                ),
            )
        )
    ).mappings().first()
    if prior_ambiguous is not None:
        # At-most-once: never auto-resend after PROCESSING/unknown.
        await _ensure_unknown_attempt(
            db,
            message_id=message_id,
            outbox_event_id=event["id"],
            error_code="AMBIGUOUS_PRIOR_ATTEMPT",
            error_message=(
                f"Prior attempt {prior_ambiguous.status}; automatic resend suppressed"
            ),
            attempt_id=(
                prior_ambiguous.id
                if prior_ambiguous.status == "PROCESSING"
                else None
            ),
        )
        return PreparedTelegramDelivery(
            action="skip_unknown",
            message_id=message_id,
            conversation_id=row.conversationId,
            chat_id=None,
            body=None,
            attempt_id=None,
            idempotency_key=event["idempotencyKey"],
        )

    identity_condition = (
        channel_identity_table.c.leadId == row.leadId
        if row.leadId
        else channel_identity_table.c.studentId == row.studentId
    )
    identity = (
        await db.execute(
            select(channel_identity_table).where(
                channel_identity_table.c.channel == "TELEGRAM",
                identity_condition,
            )
        )
    ).mappings().first()
    if identity is None:
        raise ValueError("Telegram channel identity not found")
    chat_id = (identity.metadataJson or {}).get("chat_id")
    if not chat_id:
        raise ValueError("Telegram chat id is missing")

    attempt = (
        await db.execute(
            select(func.count())
            .select_from(delivery_attempt_table)
            .where(
                delivery_attempt_table.c.messageId == message_id,
                delivery_attempt_table.c.provider == "TELEGRAM",
            )
        )
    ).scalar_one() + 1
    attempt_id = uuid4().hex
    now = datetime.now(UTC)
    await db.execute(
        insert(delivery_attempt_table).values(
            id=attempt_id,
            messageId=message_id,
            outboxEventId=event["id"],
            provider="TELEGRAM",
            attempt=attempt,
            status="PROCESSING",
            createdAt=now,
        )
    )
    return PreparedTelegramDelivery(
        action="send",
        message_id=message_id,
        conversation_id=row.conversationId,
        chat_id=str(chat_id),
        body=row.body,
        attempt_id=attempt_id,
        idempotency_key=event["idempotencyKey"],
    )


async def call_telegram_delivery(
    prepared: PreparedTelegramDelivery,
) -> TelegramCallResult:
    """HTTP send outside any DB transaction (run under lease heartbeat)."""
    if prepared.action != "send":
        raise RuntimeError("call_telegram_delivery requires action=send")
    assert prepared.chat_id is not None and prepared.body is not None
    provider_message_id = await TelegramAdapter(
        get_settings().telegram_bot_token
    ).send_text(
        prepared.chat_id,
        prepared.body,
        idempotency_key=prepared.idempotency_key,
    )
    return TelegramCallResult(provider_message_id=provider_message_id)


async def finalize_telegram_delivery(
    db: AsyncSession,
    prepared: PreparedTelegramDelivery,
    *,
    result: TelegramCallResult | None = None,
    error: Exception | None = None,
) -> Literal["success", "unknown", "failed"]:
    """Update DeliveryAttempt / message after the external call. No HTTP."""
    now = datetime.now(UTC)
    if prepared.action in {"skip_success", "skip_unknown"}:
        return "success" if prepared.action == "skip_success" else "unknown"

    if prepared.attempt_id is None:
        raise RuntimeError("finalize_telegram_delivery missing attempt_id")

    if result is not None:
        await db.execute(
            update(delivery_attempt_table)
            .where(delivery_attempt_table.c.id == prepared.attempt_id)
            .values(
                status="SUCCESS",
                providerResponseId=result.provider_message_id,
            )
        )
        await db.execute(
            update(conversation_message_table)
            .where(conversation_message_table.c.id == prepared.message_id)
            .values(
                providerMessageId=result.provider_message_id,
                deliveryStatus="SENT",
                sentAt=now,
                updatedAt=now,
            )
        )
        if prepared.conversation_id:
            await db.execute(
                update(conversation_table)
                .where(conversation_table.c.id == prepared.conversation_id)
                .values(lastOutboundAt=now, updatedAt=now)
            )
        return "success"

    if error is None:
        raise RuntimeError("finalize_telegram_delivery requires result or error")

    if _is_ambiguous_after_send_start(error):
        await _ensure_unknown_attempt(
            db,
            message_id=prepared.message_id,
            outbox_event_id="",
            error_code=type(error).__name__,
            error_message=str(error),
            attempt_id=prepared.attempt_id,
        )
        return "unknown"

    await db.execute(
        update(delivery_attempt_table)
        .where(delivery_attempt_table.c.id == prepared.attempt_id)
        .values(
            status="FAILED",
            errorCode=type(error).__name__,
            errorMessage=str(error)[:1000],
        )
    )
    return "failed"


async def deliver_telegram_message(
    db: AsyncSession, event: dict[str, Any]
) -> None:
    """Compatibility helper for in-process tests.

    Prefer prepare → call → finalize via the outbox worker so HTTP never runs
    inside the prepare transaction. This helper still sequences all three and
    must not be used when ``db`` is held open across the HTTP call in production.
    """
    prepared = await prepare_telegram_delivery(db, event)
    if prepared.action != "send":
        return
    # Flush prepare marker before HTTP when the caller shares one session.
    await db.flush()
    try:
        result = await call_telegram_delivery(prepared)
    except Exception as error:
        outcome = await finalize_telegram_delivery(db, prepared, error=error)
        if outcome == "failed":
            raise
        return
    await finalize_telegram_delivery(db, prepared, result=result)

File: src/server/delivery/telegram.ts
import type { OutboxEvent, Prisma } from "@prisma/client";
import { prisma } from "@/lib/db";

export const DELIVERY_STATUS = {
  PENDING: "PENDING",
  PROCESSING: "PROCESSING",
  SENT: "SENT",
  FAILED: "FAILED",
  UNKNOWN_REQUIRES_REVIEW: "UNKNOWN_REQUIRES_REVIEW",
} as const;

export type TelegramPrepareAction = "send" | "skip_success" | "skip_unknown";

export type PreparedTelegramDelivery = {
  action: TelegramPrepareAction;
  messageId: string;
  conversationId: string;
  chatId: string | null;
  body: string | null;
  attemptId: string | null;
  idempotencyKey: string;
};

export type TelegramCallResult = {
  providerMessageId: string;
};

export class TelegramRetryableError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "TelegramRetryableError";
  }
}

function metaChatId(metadata: Prisma.JsonValue | null | undefined): string | null {
  if (!metadata || typeof metadata !== "object" || Array.isArray(metadata)) {
    return null;
  }
  const chatId = (metadata as { chat_id?: unknown }).chat_id;
  return chatId != null ? String(chatId) : null;
}

export function isAmbiguousTelegramError(error: unknown): boolean {
  if (!(error instanceof Error)) return false;
  const name = error.name;
  if (
    name === "TimeoutError" ||
    name === "AbortError" ||
    name === "FetchError"
  ) {
    return true;
  }
  if (error.message.includes("fetch failed") || error.message.includes("network")) {
    return true;
  }
  if ("status" in error && typeof (error as { status?: number }).status === "number") {
    const status = (error as { status: number }).status;
    return status >= 500;
  }
  return false;
}

export async function prepareTelegramDelivery(
  event: OutboxEvent,
): Promise<PreparedTelegramDelivery> {
  const payload = event.payloadJson;
  if (!payload || typeof payload !== "object" || Array.isArray(payload)) {
    throw new Error("telegram.send payload must be an object");
  }
  const messageId = String(
    (payload as { messageId?: unknown }).messageId ?? "",
  );
  if (!messageId) {
    throw new Error("telegram.send requires messageId");
  }

  return prisma.$transaction(async (tx) => {
    const message = await tx.conversationMessage.findUnique({
      where: { id: messageId },
      include: { conversation: true },
    });
    if (!message) {
      throw new Error(`ConversationMessage ${messageId} not found`);
    }

    const idempotencyKey = event.idempotencyKey;

    if (
      message.deliveryStatus === DELIVERY_STATUS.SENT ||
      message.deliveryStatus === "SUCCESS"
    ) {
      return {
        action: "skip_success" as const,
        messageId,
        conversationId: message.conversationId,
        chatId: null,
        body: message.body,
        attemptId: null,
        idempotencyKey,
      };
    }

    const prior = await tx.deliveryAttempt.findMany({
      where: { messageId, provider: "TELEGRAM" },
      orderBy: { attempt: "desc" },
    });

    const hasSuccess = prior.some(
      (a) => a.status === DELIVERY_STATUS.SENT || a.status === "SUCCESS",
    );
    if (hasSuccess) {
      return {
        action: "skip_success" as const,
        messageId,
        conversationId: message.conversationId,
        chatId: null,
        body: message.body,
        attemptId: null,
        idempotencyKey,
      };
    }

    const blocking = prior.find(
      (a) =>
        a.status === DELIVERY_STATUS.PROCESSING ||
        a.status === DELIVERY_STATUS.UNKNOWN_REQUIRES_REVIEW,
    );
    if (blocking) {
      if (blocking.status === DELIVERY_STATUS.PROCESSING) {
        await tx.deliveryAttempt.update({
          where: { id: blocking.id },
          data: {
            status: DELIVERY_STATUS.UNKNOWN_REQUIRES_REVIEW,
            errorCode: "RECLAIM_NO_RESEND",
            errorMessage:
              "Prior attempt left PROCESSING; refusing automatic resend",
          },
        });
      }
      await tx.conversationMessage.update({
        where: { id: messageId },
        data: { deliveryStatus: DELIVERY_STATUS.UNKNOWN_REQUIRES_REVIEW },
      });
      return {
        action: "skip_unknown" as const,
        messageId,
        conversationId: message.conversationId,
        chatId: null,
        body: message.body,
        attemptId: blocking.id,
        idempotencyKey,
      };
    }

    const conversation = message.conversation;
    const identity = await tx.channelIdentity.findFirst({
      where: {
        channel: "TELEGRAM",
        OR: [
          ...(conversation.leadId ? [{ leadId: conversation.leadId }] : []),
          ...(conversation.studentId
            ? [{ studentId: conversation.studentId }]
            : []),
        ],
      },
      orderBy: { updatedAt: "desc" },
    });

    const chatId = metaChatId(identity?.metadataJson ?? null);
    if (!chatId) {
      throw new TelegramRetryableError(
        "No Telegram chat_id on ChannelIdentity metadata",
      );
    }
    if (!message.body?.trim()) {
      throw new Error("Outbound message body is empty");
    }

    const nextAttempt = (prior[0]?.attempt ?? 0) + 1;
    const attempt = await tx.deliveryAttempt.create({
      data: {
        messageId,
        outboxEventId: event.id,
        provider: "TELEGRAM",
        attempt: nextAttempt,
        status: DELIVERY_STATUS.PROCESSING,
      },
    });

    await tx.conversationMessage.update({
      where: { id: messageId },
      data: { deliveryStatus: DELIVERY_STATUS.PROCESSING },
    });

    return {
      action: "send" as const,
      messageId,
      conversationId: message.conversationId,
      chatId,
      body: message.body,
      attemptId: attempt.id,
      idempotencyKey,
    };
  });
}

export async function callTelegramDelivery(
  prepared: PreparedTelegramDelivery,
  env: NodeJS.ProcessEnv = process.env,
): Promise<TelegramCallResult> {
  if (prepared.action !== "send" || !prepared.chatId || !prepared.body) {
    throw new Error("callTelegramDelivery requires action=send with chatId/body");
  }
  const token = env.TELEGRAM_BOT_TOKEN?.trim();
  if (!token) {
    throw new Error("TELEGRAM_BOT_TOKEN is not configured");
  }

  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), 20_000);
  try {
    const response = await fetch(
      `https://api.telegram.org/bot${token}/sendMessage`,
      {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
          "X-IMMIGROME-Idempotency-Key": prepared.idempotencyKey,
        },
        body: JSON.stringify({
          chat_id: prepared.chatId,
          text: prepared.body,
          disable_web_page_preview: true,
        }),
        signal: controller.signal,
      },
    );

    const data = (await response.json().catch(() => null)) as {
      ok?: boolean;
      result?: { message_id?: number | string };
      description?: string;
    } | null;

    if (!response.ok) {
      const err = new Error(
        data?.description ?? `Telegram HTTP ${response.status}`,
      ) as Error & { status: number };
      err.status = response.status;
      err.name = response.status >= 500 ? "TelegramServerError" : "TelegramClientError";
      throw err;
    }
    if (!data?.ok || data.result?.message_id == null) {
      throw new Error("Telegram returned an invalid sendMessage response");
    }
    return { providerMessageId: String(data.result.message_id) };
  } finally {
    clearTimeout(timer);
  }
}

export async function finalizeTelegramDelivery(
  prepared: PreparedTelegramDelivery,
  options:
    | { result: TelegramCallResult }
    | { error: unknown }
    | { skip: "success" | "unknown" },
): Promise<"completed" | "retry"> {
  if ("skip" in options) {
    return "completed";
  }

  if ("result" in options) {
    await prisma.$transaction(async (tx) => {
      if (prepared.attemptId) {
        await tx.deliveryAttempt.update({
          where: { id: prepared.attemptId },
          data: {
            status: DELIVERY_STATUS.SENT,
            providerResponseId: options.result.providerMessageId,
            errorCode: null,
            errorMessage: null,
          },
        });
      }
      await tx.conversationMessage.update({
        where: { id: prepared.messageId },
        data: {
          deliveryStatus: DELIVERY_STATUS.SENT,
          providerMessageId: options.result.providerMessageId,
          sentAt: new Date(),
        },
      });
    });
    return "completed";
  }

  const error = options.error;
  const message = error instanceof Error ? error.message : String(error);
  const ambiguous = isAmbiguousTelegramError(error);

  if (ambiguous) {
    await prisma.$transaction(async (tx) => {
      if (prepared.attemptId) {
        await tx.deliveryAttempt.update({
          where: { id: prepared.attemptId },
          data: {
            status: DELIVERY_STATUS.UNKNOWN_REQUIRES_REVIEW,
            errorCode: "AMBIGUOUS_TRANSPORT",
            errorMessage: message.slice(0, 1000),
          },
        });
      }
      await tx.conversationMessage.update({
        where: { id: prepared.messageId },
        data: { deliveryStatus: DELIVERY_STATUS.UNKNOWN_REQUIRES_REVIEW },
      });
    });
    return "completed";
  }

  await prisma.$transaction(async (tx) => {
    if (prepared.attemptId) {
      await tx.deliveryAttempt.update({
        where: { id: prepared.attemptId },
        data: {
          status: DELIVERY_STATUS.FAILED,
          errorCode: "TELEGRAM_CLIENT_ERROR",
          errorMessage: message.slice(0, 1000),
        },
      });
    }
    await tx.conversationMessage.update({
      where: { id: prepared.messageId },
      data: { deliveryStatus: DELIVERY_STATUS.FAILED },
    });
  });
  return "retry";
}
